datafusion/common/src/tree_node.rs

```diff
@@ -204,6 +214,24 @@ pub trait TreeNode: Sized {
         apply_impl(self, &mut f)
     }
 
+    fn apply_ref<'n, F: FnMut(&'n Self) -> Result<TreeNodeRecursion>>(
```

Comment:
This API would be helpful in other areas to avoid cloning -- I am very much in favor of adding it